### PR TITLE
feat: Allow predicates to own object and evaluate against borrowed types

### PR DESCRIPTION
**Pull request by rshearman**
It is very useful to be able to dynamically construct an object and have that object owned by the predicate, yet evaluate against an unowned type related to the owned one. An obvious example is a String being owned by the predicate but being compared against &strs.

Therefore, implement Predicate for Eq/OrdPredicate that store an object that implements Borrow for the predicate type, replacing existing impls of Predicate<T> for Eq/OrdPredicate<T> and Eq/OrdPredicate<&T>. This is backwards compatible as there are blanket implementations of Borrow<T> for T and Borrow<T> for &T.

Note that Borrow imposes more requirements than are actually required and AsRef would be sufficient. However, AsRef doesn't have a blanket implementation for T and thus the existing impl of Predicate<T> for EqPredicate<T> is still required, but results in a conflict since T may also implement AsRef<T>. Requiring Borrow instead of AsRef is sufficient for common use cases though.

This addresses #20 more completely.

### CODE REVIEW

File: src/ord.rs [hunk 1]
```diff
@@ -35,47 +35,24 @@ impl fmt::Display for EqOps {
...
-            case.add_product(reflection::Product::new(
-                "var",
-                utils::DebugAdapter::new(variable).to_string(),
-            ))
-        })
-    }
-}
-
-impl<'a, T> Predicate<T> for EqPredicate<&'a T>
+impl<T, P> Predicate<T> for EqPredicate<P>
```
Comment (epage) [inline src/ord.rs:43]: imo using different generic parameters between blocks is confusing, please make them consistent (`EqPredicate<T>` vs `EqPredicate<P>`).  Ideally anything deviating from the existing would be done in its own commit before this one or in a whole PR before this one,

File: src/ord.rs [hunk 2]
```diff
@@ -35,47 +35,24 @@ impl fmt::Display for EqOps {
...
         match self.op {
-            EqOps::Equal => variable.eq(self.constant),
-            EqOps::NotEqual => variable.ne(self.constant),
+            EqOps::Equal => variable.eq(self.constant.borrow()),
+            EqOps::NotEqual => variable.ne(self.constant.borrow()),
         }
     }
 
-    fn find_case<'b>(&'b self, expected: bool, variable: &T) -> Option<reflection::Case<'b>> {
+    fn find_case<'a>(&'a self, expected: bool, variable: &T) -> Option<reflection::Case<'a>> {
```
Comment (epage) [inline src/ord.rs:55]: If we're splitting things out into separate commits, please split this one oput

File: src/ord.rs [hunk 3]
```diff
@@ -178,51 +159,26 @@ impl fmt::Display for OrdOps {
...
-            case.add_product(reflection::Product::new(
-                "var",
-                utils::DebugAdapter::new(variable).to_string(),
-            ))
-        })
-    }
-}
-
-impl<'a, T> Predicate<T> for OrdPredicate<&'a T>
+impl<T, P> Predicate<T> for OrdPredicate<P>
```
Comment (epage) [inline src/ord.rs:167]: What about the predicates outside of this file?

Comment (rshearman) [reply src/ord.rs:167]: I'm not aware of any other general purpose predicates that could benefit from the same approach, but if you have any suggestions then I can take a look. I would suggest that the specific use predicates of path and str are outside the scope of this PR though.

File: src/ord.rs (hunk 1 above)
Comment (rshearman) [reply src/ord.rs:43]: Ok, I can invert the use of predicate parameters here and below.

File: src/ord.rs (hunk 3 above)
Comment (epage) [reply src/ord.rs]: #49 (linked from #20) touched predicates in `iter.rs` as well.
